### PR TITLE
Implement initial support for intralinks.

### PR DESCRIPTION
**Pull request by orium**
Initial support for intralinks paths of the form `crate::⋯`.  I intend to add support for `std::⋯` in another PR.

### CODE REVIEW

**Comment by orium** [comment]
I have implemented support for standard library (`std`/`alloc`/`core`), but I think it is easier to open another PR once this one lands.

@phaazon  Can you take a look at this one when you have some time? :slightly_smiling_face:

**Comment by phaazon** [comment]
Hello! Sure I will!

File: src/lib.rs [hunk 1]
```diff
@@ -96,6 +98,16 @@ impl Manifest {
...
   }
 
+  /// Extract the path to the readme file from the manifest.
+  pub fn crate_name(&self) -> &str {
+    self
+      .toml
+      .get("package")
+      .and_then(|p| p.get("name"))
+      .and_then(Value::as_str)
+      .expect("Failed to get the name of the crate")
```
Comment (phaazon) [inline src/lib.rs:108]: I’m not a huge fan of this. We should simply return `Option<&str>` and let the caller decide what to do.

File: src/main.rs [hunk 2]
```diff
@@ -145,10 +152,35 @@ fn main() {
...
+            };
+            let doc = intralinks::rewrite_markdown_links(
+              &doc,
+              &symbols_type,
+              manifest.crate_name(),
+              emit_warning,
+            );
+
+            transform_readme(&readme, doc, crlf).map(|new| (readme, new))
+          });
```
Comment (phaazon) [inline src/main.rs:183]: I think the `main` function is alreadyl ong enough. Maybe you can wrap this logic in a function, or call it in `transform_readme`?

File: src/intralinks.rs [hunk 3]
```diff
@@ -0,0 +1,936 @@
...
+  let mut mod_file = dir.to_path_buf();
+  mod_file.push(format!("{}.rs", module));
+
+  if mod_file.is_file() {
+    return Some(mod_file);
+  }
+
+  let mut mod_file = dir.to_path_buf();
+  mod_file.push(module.to_string());
+  mod_file.push("mod.rs");
```
Comment (phaazon) [inline src/intralinks.rs:63]: You should be able to do that with `Path::join` instead.

**Review by phaazon — CHANGES_REQUESTED** [review]
First pass of comments. Thanks for doing this! Btw, you seem to be parsing Rust file but is it really needed? Can’t we just use the crate name for all Markdown links looking like intra-links? What is the incentive for `syn`?

**Comment by orium** [comment]
> First pass of comments. Thanks for doing this! 

My pleasure.  I'll address them soon.

> Btw, you seem to be parsing Rust file but is it really needed? Can’t we just use the crate name for all Markdown links looking like intra-links? What is the incentive for `syn`?

The only reason for using `syn` is that you need to discover what type of symbol is being references (`struct`, `trait`, `fn`, `mod`, etc...), because the link contains that (e.g. https://docs.rs/archery/0.4.0/archery/shared_pointer/struct.SharedPointer.html or https://docs.rs/archery/0.4.0/archery/shared_pointer/kind/trait.SharedPointerKind.html).

I'm being careful to to explore modules that are not referenced, so we do the minimum amount of parsing necessary.  Also we don't fail the whole thing when we can't find a symbol in the code (we emit a warning).  But currently we are failing entirely if the code fails to be parsed (which I think is reasonable).

**Comment by orium** [comment]
Addressed comments.

**Comment by phaazon** [comment]
Yeah, seems fine. That’s a massive contribution, thanks a lot! I’ll do a second and final pass of review, especially in the intralinks module.

File: src/intralinks.rs [hunk 4]
```diff
@@ -0,0 +1,933 @@
...
+  IOError(std::io::Error),
+  ParseError(syn::Error),
+  NoModuleFile(String, PathBuf),
+}
+
+impl fmt::Display for IntraLinkError {
+  fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+    match self {
+      IntraLinkError::IOError(err) => write!(f, "IO error: {}", err),
+      IntraLinkError::ParseError(err) => write!(f, "NoModuleFile error: {}", err),
```
Comment (phaazon) [inline src/intralinks.rs:21]: The error description doesn’t seem to be meaningful regarding the variant.

Comment (phaazon) [reply src/intralinks.rs:21]: (I guess you copy-pasted for the `IntraLinkError::NoModuleFile` variant.

File: src/intralinks.rs [hunk 5]
```diff
@@ -0,0 +1,933 @@
...
+  fn from(err: syn::Error) -> Self {
+    IntraLinkError::ParseError(err)
+  }
+}
+
+fn file_ast<P: AsRef<Path>>(file: P) -> Result<syn::File, IntraLinkError> {
+  let mut file = File::open(file)?;
+  let mut src = String::new();
+
+  file.read_to_string(&mut src)?;
```
Comment (phaazon) [inline src/intralinks.rs:47]: You can replace all this with:

```suggestion
  let src = fs::read_to_string()?;
```

Comment (phaazon) [reply src/intralinks.rs:47]: (i.e. you’ll need to import `std::fs`).

File: src/intralinks.rs [hunk 6]
```diff
@@ -0,0 +1,933 @@
...
+  fn new(anchor: FQIdentifierAnchor) -> FQIdentifier {
+    FQIdentifier {
+      anchor,
+      path: Vec::new(),
+    }
+  }
+
+  fn from_string(s: &str) -> Option<FQIdentifier> {
+    let anchor;
+    let str;
```
Comment (phaazon) [inline src/intralinks.rs:92]: Not a huge fan of this variable name. It’s the same symbol as the `str` type. Plus, it doesn’t help understand what this variable is for.

File: src/intralinks.rs [hunk 7]
```diff
@@ -0,0 +1,933 @@
...
+      return None;
+    }
+
+    if str.is_empty() {
+      return None;
+    }
+
+    Some(FQIdentifier {
+      anchor,
+      path: str.split("::").map(|s| s.to_owned()).collect(),
```
Comment (phaazon) [inline src/intralinks.rs:114]: You can replace the lambda with:

```suggestion
      path: str.split("::").map(str::to_owned).collect(),
```

File: src/intralinks.rs [hunk 8]
```diff
@@ -0,0 +1,933 @@
...
+    })
+  }
+
+  fn push(&self, s: &str) -> FQIdentifier {
+    let mut new = self.clone();
+    new.path.push(s.to_string());
+    new
+  }
```
Comment (phaazon) [inline src/intralinks.rs:122]: I don’t like this. It means adding element will always be _O(n)_. You can switch to `&mut self` for this and let people if they want to at call site.

File: src/intralinks.rs [hunk 9]
```diff
@@ -0,0 +1,933 @@
...
+    std::iter::successors(Some(ancestor), |ancestor| match ancestor.path.is_empty() {
+      true => None,
+      false => {
+        let mut new_ancestor = ancestor.clone();
+        new_ancestor.path.pop();
+        Some(new_ancestor)
+      }
+    })
+  }
+}
```
Comment (phaazon) [inline src/intralinks.rs:137]: This function is _O(n²)_ in allocations. :eyes:. Looking at what you want to do, I think you should be able to generate your list without allocating anything by using references instead.

**Review by phaazon — CHANGES_REQUESTED** [review]

File: src/intralinks.rs (hunk 6 above)
Comment (orium) [reply src/intralinks.rs:92]: it is hard to give it a meaningful name, it is slightly better now.

File: src/intralinks.rs (hunk 9 above)
Comment (orium) [reply src/intralinks.rs:137]: I can't think of a way except to either add a lifetime to `FQIdentifier` or to use an `Rc` and share the `Vec<_>`.  We still will need to do O(n) allocations.

Comment (orium) [reply src/intralinks.rs]: Implemented with `Rc<_>`.  I'm not convinced this is worth the extra code complexity, particularly when taking into consideration that most symbol paths will have a relatively small depth (and we only do this for the symbols mentioned in the `//!` doc).

Comment (phaazon) [reply src/intralinks.rs]: Hm. That seems like a complex solution but maybe I had underestimated the problem. I’ll try to look a bit more into it and I guess we can merge this.

File: src/lib.rs [hunk 10]
```diff
@@ -238,9 +251,9 @@ where
   transform_inner_doc(&doc, show_hidden_doc, crlf)
 }
 
-#[derive(Debug, Eq, PartialEq)]
```
Comment (phaazon) [inline src/lib.rs:241]: Can you leave `Debug` there? It’s implemented for all types in your variant.

**Comment by phaazon** [comment]
Okay, sorry for the delay. I’ll have a final look tonight and merge / release it. Thanks again for your patience. :sweat_smile:

File: src/lib.rs [hunk 11]
```diff
@@ -271,17 +294,27 @@ where
...
+  entry_point: &Path,
+  emit_warning: impl FnMut(&str),
+) -> Result<String, TransformError> {
+  let symbols: HashSet<FQIdentifier> = intralinks::extract_markdown_intralink_symbols(doc);
+  let modules = intralinks::all_supermodules(symbols.iter());
+  let symbols_type =
+    intralinks::crate_symbols_type(&entry_point, |module| modules.contains(module))?;
+
+  Ok(intralinks::rewrite_markdown_links(
+    &doc,
```
Comment (phaazon) [inline src/lib.rs:309]: That reference shouldn’t be needed.

**Comment by orium** [comment]
> Okay, sorry for the delay. I’ll have a final look tonight and merge / release it. Thanks again for your patience. sweat_smile

No worries.  Don't release it yet because I have support for links to stdlib ready in a branch.  I can open the PR once this lands.  It is way smaller, so it shouldn't be much easier to review than this PR :slightly_smiling_face:

**Review by phaazon — APPROVED** [review]
Let’s go! :) Thanks again!

**Comment by aldanor** [comment]
Is it in plans to release a current version with this? (as it looks immensely helpful as of right now)

**Comment by orium** [comment]
I have PR #38 open to also allow you to have links to the standard library.  Makes sense to include that in the release.

**Comment by phaazon** [comment]
I’m currently reviewing all of the recently merged PRs. I’ll be doing a release later in the night.

File: src/intralinks.rs [hunk 12]
```diff
@@ -0,0 +1,960 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+use syn::export::fmt::Display;
+use syn::export::{Debug, Formatter};
```
Comment (phaazon) [inline src/intralinks.rs:7]: Goddammit I wish I saw that earlier…

**Comment by phaazon** [comment]
So after a bad surprise, I have decided not to release anything until I have fixed the current situation. A regression was introduced by #35, breaking the build because of a hidden module (`serde::export`). I’ll clean up everything tomorrow.

**Pull request merged**